### PR TITLE
Make exception handler effectful

### PR DESCRIPTION
**Pull request by danielleontiev**
This changes signature of `ExceptionHandler` from `ExceptionContext => Option[ValuedEndpointOutput[_]]` to `ExceptionContext => F[ValuedEndpointOutput[_]]`. It allows to handle errors effectfully - now it's requires to completely redefine `ExceptionInterceptor` to do so.

If this change makes sense I think I can also create PR to make RejectionHandler effectful too.

### CODE REVIEW

File: server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
 import sttp.model.StatusCode
-import sttp.tapir.{server, _}
+import sttp.monad.MonadError
+import sttp.tapir._
 import sttp.tapir.server.model.ValuedEndpointOutput
 
-trait ExceptionHandler {
-  def apply(ctx: ExceptionContext): Option[ValuedEndpointOutput[_]]
+trait ExceptionHandler[F[_]] {
+  def apply(ctx: ExceptionContext)(implicit monad: MonadError[F]): F[ValuedEndpointOutput[_]]
```
Comment (adamw) [inline server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala:9]: I think I would retain the possibility to handle only some exceptions, hence returning `F[Option[ValuedEndpointOutput[_]]]`

File: server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala [hunk 2]
```diff
@@ -1,19 +1,20 @@
...
 
 import sttp.model.StatusCode
-import sttp.tapir.{server, _}
+import sttp.monad.MonadError
+import sttp.tapir._
 import sttp.tapir.server.model.ValuedEndpointOutput
 
-trait ExceptionHandler {
-  def apply(ctx: ExceptionContext): Option[ValuedEndpointOutput[_]]
+trait ExceptionHandler[F[_]] {
```
Comment (adamw) [inline server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala:8]: maybe we can add a method to conveniently create pure handlers, that is given `ExceptionContext => Option[VEO[_]]`

File: server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala (hunk 1 above)
Comment (danielleontiev) [reply server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala:9]: Yes, you are right. It's possible with rising error from handler, but returning None is more convenient

File: server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala (hunk 2 above)
Comment (danielleontiev) [reply server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala:8]: I like the idea but I struggle to create descriptive name for it. I think we can mimic cats effect convention and rename apply to `handleError` and name second method `handleErrorWith`. What do you think?

Comment (adamw) [reply server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala:8]: maybe simply `object ExceptionHandler { def pure(...) ... }` ?

Comment (danielleontiev) [reply server/core/src/main/scala/sttp/tapir/server/interceptor/exception/ExceptionHandler.scala:8]: Yes, I've finally decided to use `ExceptionHandler` comainion object, already in progress :)

**Comment by adamw** [comment]
Thanks!

**Pull request merged**